content/300-guides/100-performance-and-optimization/090-metrics.mdx

```diff
@@ -0,0 +1,416 @@
+    statsd.gauge('prisma.' + counter.key, counter.value, (...res) => {})
+  })
+
+  metrics.gauges.forEach((counter: any) => {
+    statsd.gauge('prisma.' + counter.key, counter.value, (...res) => {})
+  })
+
+  if (previousHistograms === null) {
+    previousHistograms = diffMetrics(metrics.histograms)
+    return
+  }
+
+  let diffHistograms = diffMetrics(metrics.histograms)
+
+  diffHistograms.forEach((diffHistograms: any, histogramIndex: any) => {
+    diffHistograms.value.buckets.forEach((values: any, bucketIndex: any) => {
+      let [bucket, count] = values
+      let [_, prev] =
+        previousHistograms[histogramIndex].value.buckets[bucketIndex]
+      let change = count - prev
+
+      for (let sendTimes = 0; sendTimes < change; sendTimes++) {
+        statsd.timing('prisma.' + diffHistograms.key, bucket)
+      }
+    })
+  })
+
+  previousHistograms = diffHistograms
+}
+
+setInterval(async () => await statsdSender(), 10000)
+```
+
+## Retrieve metrics in Prometheus format
+
+When you retrieve Prisma metrics in Prometheus format, you can use them in the format they are returned, or [send them to the Prometheus metrics system](#use-prisma-metrics-with-the-prometheus-metrics-system) to visualize how they change over time.
+
+To retrieve Metrics in Prometheus format, add the following lines to your application code:
+
+```ts
+let metrics = prisma.$metrics.prometheus()
+console.log(metrics)
+```
+
+This returns metrics as follows:
+
+```c
+# HELP query_total_operations The total number of operations executed
+
+# TYPE query_total_operations counter
+
+query_total_operations 0
+
+# HELP query_total_queries The total number of queries executed
+
+# TYPE query_total_queries counter
+
+query_total_queries 0
+
+# HELP pool_active_connections The number of active connections in use
+
+# TYPE pool_active_connections gauge
+
+pool_active_connections 0
+
+# HELP pool_idle_connections The number of connections that are not being used
+
+# TYPE pool_idle_connections gauge
+
+pool_idle_connections 21
+
+# HELP pool_wait_count The number of workers waiting for a connection
+
+# TYPE pool_wait_count gauge
+
+pool_wait_count
+
+# HELP query_active_transactions The number of active transactions
+
+# TYPE query_active_transactions gauge
+
+query_active_transactions
+
+# HELP pool_wait_duration_ms The wait time for a worker to get a connection.
+
+# TYPE pool_wait_duration_ms histogram
+
+pool_wait_duration_ms_bucket{le="0"} 0
+pool_wait_duration_ms_bucket{le="1"} 1
+pool_wait_duration_ms_bucket{le="5"} 1
+pool_wait_duration_ms_bucket{le="10"} 1
+pool_wait_duration_ms_bucket{le="50"} 1
+pool_wait_duration_ms_bucket{le="100"} 1
+pool_wait_duration_ms_bucket{le="500"} 1
+pool_wait_duration_ms_bucket{le="1000"} 1
+pool_wait_duration_ms_bucket{le="5000"} 1
+pool_wait_duration_ms_bucket{le="50000"} 1
+pool_wait_duration_ms_bucket{le="+Inf"} 1
+pool_wait_duration_ms_sum 0.050875000000000004
+pool_wait_duration_ms_count 1
```

Comment:
Here is an actual sample which takes care of whitespace issues and adds missing histograms

```suggestion
    # HELP query_total_operations
    # TYPE query_total_operations counter
    query_total_operations 2

    # HELP query_total_queries
    # TYPE query_total_queries counter
    query_total_queries 5

    # HELP pool_active_connections The number of active connections in use.
    # TYPE pool_active_connections gauge
    pool_active_connections 0

    # HELP pool_idle_connections The number of connections that are not being used
    # TYPE pool_idle_connections gauge
    pool_idle_connections 21

    # HELP pool_wait_count The number of workers waiting for a connection.
    # TYPE pool_wait_count gauge
    pool_wait_count 0

    # HELP query_active_transactions
    # TYPE query_active_transactions gauge
    query_active_transactions 0

    # HELP pool_wait_duration_ms The wait time for a worker to get a connection.
    # TYPE pool_wait_duration_ms histogram
    pool_wait_duration_ms_bucket{le="0"} 0
    pool_wait_duration_ms_bucket{le="1"} 3
    pool_wait_duration_ms_bucket{le="5"} 3
    pool_wait_duration_ms_bucket{le="10"} 3
    pool_wait_duration_ms_bucket{le="50"} 3
    pool_wait_duration_ms_bucket{le="100"} 3
    pool_wait_duration_ms_bucket{le="500"} 3
    pool_wait_duration_ms_bucket{le="1000"} 3
    pool_wait_duration_ms_bucket{le="5000"} 3
    pool_wait_duration_ms_bucket{le="50000"} 3
    pool_wait_duration_ms_bucket{le="+Inf"} 3
    pool_wait_duration_ms_sum 0.023208
    pool_wait_duration_ms_count 3

    # HELP query_operation_total_elapsed_time_ms
    # TYPE query_operation_total_elapsed_time_ms histogram
    query_operation_total_elapsed_time_ms_bucket{le="0"} 0
    query_operation_total_elapsed_time_ms_bucket{le="1"} 1
    query_operation_total_elapsed_time_ms_bucket{le="5"} 2
    query_operation_total_elapsed_time_ms_bucket{le="10"} 2
    query_operation_total_elapsed_time_ms_bucket{le="50"} 2
    query_operation_total_elapsed_time_ms_bucket{le="100"} 2
    query_operation_total_elapsed_time_ms_bucket{le="500"} 2
    query_operation_total_elapsed_time_ms_bucket{le="1000"} 2
    query_operation_total_elapsed_time_ms_bucket{le="5000"} 2
    query_operation_total_elapsed_time_ms_bucket{le="50000"} 2
    query_operation_total_elapsed_time_ms_bucket{le="+Inf"} 2
    query_operation_total_elapsed_time_ms_sum 3.197624
    query_operation_total_elapsed_time_ms_count 2

    # HELP query_total_elapsed_time_ms
    # TYPE query_total_elapsed_time_ms histogram
    query_total_elapsed_time_ms_bucket{le="0"} 0
    query_total_elapsed_time_ms_bucket{le="1"} 5
    query_total_elapsed_time_ms_bucket{le="5"} 5
    query_total_elapsed_time_ms_bucket{le="10"} 5
    query_total_elapsed_time_ms_bucket{le="50"} 5
    query_total_elapsed_time_ms_bucket{le="100"} 5
    query_total_elapsed_time_ms_bucket{le="500"} 5
    query_total_elapsed_time_ms_bucket{le="1000"} 5
    query_total_elapsed_time_ms_bucket{le="5000"} 5
    query_total_elapsed_time_ms_bucket{le="50000"} 5
    query_total_elapsed_time_ms_bucket{le="+Inf"} 5
    query_total_elapsed_time_ms_sum 1.8407059999999997
    query_total_elapsed_time_ms_count 5
```
```